test/tests/functional/pbs_qrun.py

```diff
@@ -92,3 +92,48 @@ def test_invalid_host_val(self):
         msg = "Server is not up"
         self.assertTrue(self.server.isUp(), msg)
         self.logger.info("As expected server is up and running")
+
+    def test_qrun_hangs(self):
+        """
+        This test submit 500 jobs with differnt equivalence class,
+        turn of scheduling and qrun job to
+        verify whether qrun hangs.
+        """
+        node = self.mom.shortname
+        err_msg = 'qrun: Not enough free nodes available'
+        joblist = []
+        self.server.manager(MGR_CMD_SET, SCHED,
+                            {'scheduling': 'False'})
+        self.server.manager(MGR_CMD_SET, NODE,
+                            {'resources_available.ncpus': 1}, id=node)
+        for walltime in range(1, 501):
+            j = Job(TEST_USER)
+            a = {'Resource_List.walltime': walltime}
+            j.set_attributes(a)
+            jid = self.server.submit(j)
+            joblist.append(jid)
+        self.logger.info("Submitted 500 jobs with different walltime")
+
+        self.server.manager(MGR_CMD_SET, SCHED,
+                            {'scheduling': 'True'})
+        self.server.manager(MGR_CMD_SET, SCHED,
+                            {'scheduling': 'False'})
+        jid = joblist[10]  # it scheduled a few jobs
+        now = int(time.time())
+        try:
+            self.server.runjob(jobid=jid)
```

Comment:
runjob() is a blocking call by default, so if qrun actually gets hung, your test will just remain hung and timeout. So, I'd recommend that you set runjob()'s async argument to True while calling it, and then all that you need to do is check that a sched cycle was started by the call, which you might want to do inside a 'finally' block as you should check for that regardless of PbsRunError being thrown.

Reply:
Thanks @agrawalravi90 for the review. 
Using "async" option will not help in this case. Below are the explanation. Let me know if am missing anything.
As per the doc:
qrun -a jobid : The qrun command exits before the job actually starts execution.
I believe in PTL "self.server.runjob(jobid=jid,async=True)" does the same thing.

If the job is unable to run and scheduling is also disabled, then irrespective of "async" parameter, it will hung. And which will turn runjob() call to blocking (unless its time out).

Reply:
Oh, ok, thanks for explaining. How about calling runjob in a separate Python thread/process? the main thread can then go ahead and check sched logs while the child thread is doing runjob, that way even if it gets hung the test will do its job correctly.

A test that checks for a command getting hung should not get hung itself, it should be able to fail/pass depending on whether the command got hung or not.

Reply:
Sure.  I will fork another process for runjob().